tensorflow/core/util/gpu_device_functions.h

```diff
@@ -725,6 +725,17 @@ __device__ inline Eigen::half GpuAtomicAdd(Eigen::half* ptr,
       ptr, [value](Eigen::half a) { return a + value; });
 }
 
+__device__ inline Eigen::bfloat16 GpuAtomicAdd(Eigen::bfloat16* ptr,
+                                               Eigen::bfloat16 value) {
+  Eigen::half ret = detail::GpuAtomicCasHelper(
+      reinterpret_cast<Eigen::half*>(ptr), [value](Eigen::half a) {
+        Eigen::bfloat16 ret =
+            Eigen::numext::bit_cast<Eigen::bfloat16>(a) + value;
+        return Eigen::numext::bit_cast<Eigen::half>(ret);
+      });
```

Comment:
So I assume we can't do `GpuAtomicCasHelper` with bfloat16 which is the reason for the extra conversion?

Reply:
Yes, GpuAtomicCasHelper only has the instance for fp16. Instantiating it for bf16 is not trivial so I am thinking about reusing the fp16 instance since both fp16 and bf16 have 16-bit float and the logic of atomic cas for them is same.